src/TestRunner.lua

```diff
@@ -55,30 +48,27 @@ function TestRunner.runPlanNode(session, planNode, lifecycleHooks)
 
 		messagePrefix = messagePrefix or ""
 
-		local testEnvironment = getfenv(callback)
-
-		for key, value in pairs(TestRunner.environment) do
+		local originalEnvironment = getfenv(callback)
+		local testEnvironment = setmetatable({}, { __index = originalEnvironment })
+		for key, value in pairs(planNode.environment) do
 			testEnvironment[key] = value
 		end
+		setfenv(callback, testEnvironment)
 
-		testEnvironment.fail = function(message)
-			if message == nil then
-				message = "fail() was called."
-			end
+		local nodeSuccess, nodeResult = xpcall(callback, function(message)
+			return debug.traceback(message, 2)
+		end)
 
+		if planNode.errorMessage then
```

Comment:
When is this value set? Should this error be reported and propagated before we run this test instead of here, after we've already tried to run the test callback?

Reply:
Hmm... I may need to check on this. This is supposed to get set when calling fail from within the test, but I'm not seeing that code in the PR.

Reply:
Should `fail` be mutating the test plan? I don't think we re-use test plans right now, and maybe we shouldn't, but I'm not sure if that's actually codified anywhere.

Reply:
Good point, but if the goal is to move all the injected functions into the plan environment, then how else would fail communicate with the runner? Or do you think that isn't the right goal?

Reply:
I've removed this bit for now. I'll address it in a separate CL if it needs addressing.